### PR TITLE
Add SHACL constraints and CI for checking them

### PR DESCRIPTION
**Pull request by AleksanderZawisza**
Adding shacl_constraints.ttl and adding new job to the workflow.

### CODE REVIEW

**Comment by Szwagi3r** [comment]
Also, as we wrote on teams, two different concepts (C18115, C5616) have identical prefLabel. If that was changed we could detect cycles by using sh:oneOrMorePath and recursively analyse broader concepts to compare them by prefLabels (we didn't find a way to do this directly, via IRIs) instead of manually checking if there are any concepts 9 broaders away from the focus node.

**Comment by Ostrzyciel** [comment]
> Also, as we wrote on teams, two different concepts (C18115, C5616) have identical prefLabel. If that was changed we could detect cycles by using sh:oneOrMorePath and recursively analyse broader concepts to compare them by prefLabels (we didn't find a way to do this directly, via IRIs) instead of manually checking if there are any concepts 9 broaders away from the focus node.

That's an interesting find, please use the occasion to edit one of the concepts in question to resolve this. Then you can update the SHACL rule as you suggested.

File: ontology/shacl_constraints.ttl [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@prefix exData: <http://datashapes.org/shasf/tests/expression/advanced.test.data#> .
+@prefix owl: <http://www.w3.org/2002/07/owl#> .
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
+@prefix sh: <http://www.w3.org/ns/shacl#> .
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
+<http://datashapes.org/shasf/tests/expression/advanced.test.shacl>
+    rdf:type owl:Ontology ;
+    rdfs:label "Test of advanced features" ;
```
Comment (Ostrzyciel) [inline ontology/shacl_constraints.ttl:12]: Test? :)

File: ontology/shacl_constraints.ttl [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		]
+	) .
+
+
+ex:NoCyclesRule
+	a sh:NodeShape ;
+	sh:targetClass skos:Concept  ;
+    sh:message "Path longer than 9 detected" ;
+	sh:property [
+		sh:path  (skos:broader skos:broader skos:broader skos:broader skos:broader skos:broader skos:broader skos:broader skos:broader) ;
```
Comment (Ostrzyciel) [inline ontology/shacl_constraints.ttl:59]: Please change this as you suggested in the PR.

File: ontology/shacl_constraints.ttl [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		[
+			sh:property [
+				sh:path skos:broader ;
+				sh:minCount 1 ;
+			]
+		]
+		[
+			sh:property [
+		        sh:path skos:prefLabel ;
+		        sh:pattern "^Computer science$" ;
```
Comment (Ostrzyciel) [inline ontology/shacl_constraints.ttl:48]: Why match by string? The name of the C1 concept can and will change. Can you use sh:node instead? https://www.w3.org/TR/shacl/#NodeConstraintComponent

File: ontology/shacl_constraints.ttl [hunk 4]
```diff
@@ -0,0 +1,61 @@
+@prefix ex: <http://datashapes.org/shasf/tests/expression/advanced.test.shacl#> .
+@prefix exOnt: <http://datashapes.org/shasf/tests/expression/advanced.test.ont#> .
+@prefix exData: <http://datashapes.org/shasf/tests/expression/advanced.test.data#> .
```
Comment (Ostrzyciel) [inline ontology/shacl_constraints.ttl:3]: I think there are some superfluous prefixes – please remove them.

File: ontology/shacl_constraints.ttl [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+@prefix owl: <http://www.w3.org/2002/07/owl#> .
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
+@prefix sh: <http://www.w3.org/ns/shacl#> .
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
+<http://datashapes.org/shasf/tests/expression/advanced.test.shacl>
```
Comment (Ostrzyciel) [inline ontology/shacl_constraints.ttl:10]: Please move this to some prefix that makes sense for the OpenCS ontology. Check what are the other prefixes used in OpenCS.

File: README.md [hunk 6]
```diff
@@ -14,6 +14,7 @@ See also **[the documentation](https://github.com/OpenCS-ontology/OpenCS/wiki/)*
   - `core` – core of the ontology, split into one Turtle file per concept. The files are grouped by 1000 into directories to help with potential filesystem or file browser issues.
   - `makg_extra.ttl.gz` – extra assertions carried over from MAKG
   - `opencs_to_makg.ttl.gz` – equivalence relations between OpenCS and MAKG
+  - `shacl_constraints.ttl` - the file with SHACL constraints for ontology validation
```
Comment (Ostrzyciel) [inline README.md:17]:
```suggestion
  - `shacl_constraints.ttl` – the file with SHACL constraints for ontology validation
```

Comment (Ostrzyciel) [reply README.md:17]: (nitpicking: there's a tiny difference between a regular dash and an "en dash", "półpauza" in Polish)

**Review by Ostrzyciel — CHANGES_REQUESTED** [review]

**Comment by AleksanderZawisza** [comment]
We have applied requested changes and upadated the C18115.ttl.

**Comment by AleksanderZawisza** [comment]
Also, we have found another possible mistake in the ontology: C39038.ttl and C45733.ttl are the same concept (they have different labels due to misspelling). They have the same broader concept. We did not want to remove any of them yet without consultations.

**Comment by AleksanderZawisza** [comment]
links for faster access:
https://github.com/OpenCS-ontology/OpenCS/blob/main/ontology/core/39/C39038.ttl
https://github.com/OpenCS-ontology/OpenCS/blob/main/ontology/core/45/C45733.ttl

**Comment by Ostrzyciel** [comment]
> Also, we have found another possible mistake in the ontology: C39038.ttl and C45733.ttl are the same concept (they have different labels due to misspelling). They have the same broader concept. We did not want to remove any of them yet without consultations.

Good find! Does this break your validation somehow? If not, please just submit an issue to this repository to remove either of these.

**Comment by AleksanderZawisza** [comment]
This does not break validation. We will open the issue right away.

**Review by Ostrzyciel — APPROVED** [review]

**Comment by Ostrzyciel** [comment]
> This does not break validation. We will open the issue right away.

Perfect! Very good work then. Remember to mention what you discovered in the report (you can send me an updated version of it).

**Pull request merged**